src/query.cpp

```diff
@@ -0,0 +1,226 @@
+#include "query.hpp"
+
+#include <algorithm>
+
+#include <fmt/format.h>
+#include <nlohmann/json.hpp>
+
+namespace pisa {
+
+QueryRequest::QueryRequest(QueryContainer const& data) : m_threshold(data.threshold())
+{
+    if (auto term_ids = data.term_ids(); term_ids) {
+        m_term_ids = *term_ids;
+        std::sort(m_term_ids.begin(), m_term_ids.end());
+        auto last = std::unique(m_term_ids.begin(), m_term_ids.end());
+        m_term_ids.erase(last, m_term_ids.end());
+    }
+    throw std::domain_error("Query not parsed.");
+}
+
+auto QueryRequest::term_ids() const -> gsl::span<std::uint32_t const>
+{
+    return gsl::span<std::uint32_t const>(m_term_ids);
+}
+
+auto QueryRequest::threshold() const -> std::optional<float>
+{
+    return m_threshold;
+}
+
+struct QueryContainerInner {
```

Comment:
Could we (should we) include whether the query must be `safe` as well? Or somewhere else?

Reply:
That's a good question. I would lean towards it being part of the other class: `QueryRequest`. Here's why: I can imagine having the same identical query "containers" in a JSON file, and decide to run it safely and unsafely. It's more a property of a request than a box with query data. Do I make sense?

Reply:
Yep, agreed.